### DISCUSSION TITLE
What are you working on?

### DISCUSSION THREAD

**Discussion by pragma37** [bnpr/Malt#47] [Showcase]
Let's share our Malt related art here!

Everything is welcome, from your work in progress experiments to finished animations.

**Comment by pragma37** [comment]
I'll break the ice!

I recently was asked if it's possible to get outlines on shadows (Moebius style) with Malt. 
And yep, of course you can! At first I thought you would need a post-process shader, but then I figured you can actually do it from the mesh shader too!

![moebius_shading](https://user-images.githubusercontent.com/36610999/105390793-b83ee700-5c19-11eb-9200-eeebe94e00ac.png)

The shader:
https://gist.github.com/pragma37/a0fc10cf61c7e3d26beac5e0afc12fc6

**Reply by eballeste** [reply]
super cool! for anybody that is trying to run this shader, the `NPR_lit_surface` function requires two new parameters:

```c++
NPR_lit_surface(S.position, S.normal, S.id, L, i, Settings.Receive_Shadow, Settings.Self_Shadow);
```

**Comment by markjoshua12** [comment]
Might be an unconventional use of Malt, but I've been using it for shader experiments!
It's really great to have the ability to use fragment shaders in Blender 😄

My first tests in Malt, basic raymarcher and mandelbulb
![malt-screenshot-2](https://user-images.githubusercontent.com/48671765/105530637-025fbf80-5d23-11eb-8b78-43d708632f10.png)

Trying out a volumetric shader
![malt-screenshot-1](https://user-images.githubusercontent.com/48671765/105530535-e52af100-5d22-11eb-8cec-edb779918b0e.png)

Basic raytracer ( lighting's broken atm 😛 )
![malt-screenshot-3](https://user-images.githubusercontent.com/48671765/105530911-7c904400-5d23-11eb-9f90-ca41c97f5fb4.png)

**Reply by pragma37** [reply]
That's so cool! Thank you for sharing!
Ray marching is one of those things I always want to experiment with, but I never find the right time for it.

Actually, ShaderToy inside Blender has always been one of my goals for Malt. My hope is to make an environment where graphics programmers feel at home too, so artists can benefit directly from their experiments.

**Comment by Kolupsy** [comment]
https://user-images.githubusercontent.com/88381426/143779108-fb95b9d5-8824-46be-9e54-434b4dfbcc11.mp4

Made a little test shot using Malt basically as a replacement for Eevee shading, prototyped most of the shaders in Eevee first, then basically tried to match it in Malt as closely as possible and potentially even better. 

TODO:
- fix outline of glass apprearing on steam effect (no idea how to do that yet)
- maybe add fake refraction effect if even possible to glass
- add more props / balance colors in scene
- convert some nodes into custom shader functions

**Reply by pragma37** [reply]
Looks great!
> fix outline of glass apprearing on steam effect (no idea how to do that yet)

Could you upload a minimal reproducible example?

**Comment by Kolupsy** [comment]
Yeah so its about the steam getting an outline from the glass. In general a question would be how to control the line art properly. I think i now understood the Line Simple function but other than that I often just keep dragging sliders aimlessly :D

Image:
![image](https://user-images.githubusercontent.com/88381426/144112404-79b34bde-8a4b-4513-925a-2f353a3083ad.png)

File:
https://drive.google.com/file/d/1QF75lKw-Zt2UEV_cWiZ-8hXnsBoiSgvC/view?usp=sharing

**Reply by pragma37** [reply]
I see, this is a current limitation.
The default Malt pipeline uses Depth Peeling for transparency.
https://developer.download.nvidia.com/assets/gamedev/docs/OrderIndependentTransparency.pdf
It basically consists of rendering transparent objects as opaque, but in multiple layers.
It avoids depth sorting issues and also allows applying arbitrary post-processing/compositing to each transparent layer.
But at the moment the line detection is not clever enough to take multiple layers into account.

**Comment by Kolupsy** [comment]
Very gimmicky but I did create an application template for Malt in Blender. It uses a custom startup file with new workspaces and all the shader editors replaced for malt tree editors. It also has its own preferences with malt enabled so that it does not need to load every time I open my default blender. I compiled a script that runs the application template into an executable. 

![image](https://user-images.githubusercontent.com/88381426/151776125-2573c7c8-dc36-4e47-b514-2fb74a7e7719.png)

Here is the code that I used for the executable:
```
from subprocess import Popen
blender_path = r"C:\Program Files\Blender Foundation\Blender 3.0\blender.exe"
blender_args = [
    blender_path,
    '--app-template',
    'MALT'
]
Popen( blender_args )
```

**Reply by pragma37** [reply]
Woah! That¡s nice!
I need to look into Blender templates. I updated my startup file for Malt, but that seems much better.

**Reply by Kolupsy** [reply]
Almost forgot. Most importantly of course:
Custom splash screen! :D
![image](https://user-images.githubusercontent.com/88381426/151848902-2776edcf-d767-41e9-9b3f-b7775edb07b5.png)

App templates as I see and read it are not very advanced and useful at the moment. Its rather inconvenient to install them and the installation method - parallel to how you are able to load addons from a folder separate to the app data folder - does not even work at the moment. But I was intruiged to find out whats going on with them when they recently mentioned that the devs will work on that this year. In the end it boils down to being able to launch blender with a separate startup file, user preferences and being able to execute code on startup. I havent found a great example yet but I hope this will change soon as more people are hipefully going to explore this topic.

**Comment by panthavma** [comment]
I rewrote my shader from Kronian Titans into Malt, now I can do lookdev exactly like it will appear in the game!
I'll be able to improve it faster too, and as a bonus I can finally check the palettes from here!

![image](https://user-images.githubusercontent.com/22663538/164386341-fdbe9ecd-7cc5-46f2-a03b-7fc640771526.png)
![image](https://user-images.githubusercontent.com/22663538/164386361-79b852ed-9f48-4dea-8e80-40620b66245d.png)

**Comment by Kolupsy** [comment]
I absolutely love the Malt lineart. 
For this project I wished if I had access to the world matrices of specific objects, still trying to figure this one out. Its really tough because there is pretty much no way to 'drive' malt properties in node trees. I think I tried every possible way but still no luck. Had to resort to drivers in material overrides. Not ideal for several resons but it worked. 

https://user-images.githubusercontent.com/88381426/174877344-d61def24-ca61-4b60-8730-d7e02a269f72.mp4

![viewport_wireframe](https://user-images.githubusercontent.com/88381426/174877712-aadd0861-5308-48aa-880d-85d8d38516d5.png)

**Reply by Psi25** [reply]
Did you achieve the line sketch using screen space UV in the material or is it a post process shader?

**Comment by jollyjolms** [comment]
First time using Malt! Am I late to the game Haha. I see Blender now release a NPR prototype build
![Jabari New Edyn Final_2](https://github.com/user-attachments/assets/2cf8343e-c4ed-4ede-a302-a18fe60bc8de)

 I thoroughly overloaded the scene so certainly need more optimized workflows in the future

**Comment by negicless** [comment]
ello everyone!

I’ve been using Malt and exploring different solutions with it for quite some time now.
I really enjoy everything about it, and I truly hope it continues to evolve in the future.

Personally, I focus on anime / hand-drawn style NPR rendering — it’s such a fun and expressive approach.
Here’s one of my recent mini projects, “Warm Kitchen.”

Hope you enjoy it! 😊

<img width="1500" height="1500" alt="WarmKitchen_A" src="https://github.com/user-attachments/assets/35871fee-c39d-4ae8-a104-61957e39a1f9" />
https://imgur.com/a/0vkHinc

I also shared this on Blender Artists, hoping more people will give Malt a try — even though it can be a bit tricky to learn at first, it’s absolutely worth it!
[https://blenderartists.org/t/warm-kitchen-malt-npr-test-scene/1615807](url)

**Reply by Kolupsy** [reply]
This looks super cozy. Great job!

**Reply by negicless** [reply]
> This looks super cozy. Great job!

Thanks! I am also using your MaltShadingEssentials, really good plugin for malt!

**Reply by Kolupsy** [reply]
That's nice to hear! Though i thought we merged the nodes from that project into malt directly, making it obsolete.

**Reply by pragma37** [reply]
@negicless  Looks really nice, thanks for sharing.

> I really enjoy everything about it, and I truly hope it continues to evolve in the future.

I hope that the Blender NPR support becomes good enough that most people won't even care about Malt anymore. 😄
We'll see.


Hi @Kolupsy! It's been a while!
Hope everything is going well.

And yep, all those nodes are still there, thank you again for your help.

**Reply by Kolupsy** [reply]
Hi @pragma37, I am doing very well. I hope the same goes for you. It's been some wild couple years but a couple weeks ago I got to talking about malt with @panthavma and AversionOfReality. I'm super excited to see how NPR in Blender will develop.

**Reply by panthavma** [reply]
Yeah @pragma37 I gotta thank you, Malt has been really useful for me these last few years! I'm probably one of the very few that enjoy it mainly for the ability to make my own renderer, but it's been the backbone of my current projects!

**Comment by paulbrzeski** [comment]
Hey everyone!

I'm working on a 1990s style cartoon using Malt. Currently in the production stage with our first test clip.

Work in Progress clip here: https://www.youtube.com/watch?v=2I0XrnPJ-lA

<img width="1920" height="1080" alt="0019" src="https://github.com/user-attachments/assets/345bd033-99ec-46fa-82dc-5daa66280edf" />

<details>
  <summary>Additional Screenshots</summary>
<img width="1920" height="1080" alt="Establishing shot v2" src="https://github.com/user-attachments/assets/6af73119-9590-4c16-bee3-b85d99c787b4" />
<img width="1920" height="1080" alt="0062" src="https://github.com/user-attachments/assets/5d341cf0-25fb-4f8e-80f9-faa537681aaf" />
<img width="3840" height="2092" alt="Screenshot_20251113_145520" src="https://github.com/user-attachments/assets/baf15377-740c-4d95-ad7a-f45a51bf192c" />
<img width="3840" height="2069" alt="Screenshot_20251113_145729" src="https://github.com/user-attachments/assets/bbb57b63-aa84-4a16-aa41-f610f2170e33" />
</details>

Malt has been an absolute joy to work with, I was able to achieve the look I had in my mind simply by referencing the shader materials example file and building upon that to achieve effects like thick character outlines with line boil.

There's other tricks going on under the hood to achieve the effect, full props to my partner Garrett who implemented a special face rig for the characters eyes so that they're always facing the camera and overlapping each other in that classical cartoon way that regular 3D geometry just can't do.

I'm planning to share how some of the tricks and techniques for this in a tutorial on my studio's blog, I think the world needs to know about NPR; a lot of people would be interested in learning Malt (and Blender) if they knew what was possible and how accessible it actually is.

Cheers!

**Comment by enderprism** [comment]
i finally got malt to work on blender 5.0 (i'm on arch linux, so the python version is 3.13. malt was looking for `BlenderMalt/.MaltPath/Malt/.Dependencies-313` but the directory was named `.Dependencies-311`)
i'm using Malt to rebuild my 4.4-npr-prototype NPR pipeline in 5.0 while i wait for Blender's [NPR Project](https://code.blender.org/2025/05/npr-project/) to become a thing.

`npr-prototype` version:
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/f3790600-4ba1-418d-8c12-6ba10b6d444e" />


Malt version:
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/7907e32b-7d35-42be-9513-b5d41cdc3ec9" />

**Reply by enderprism** [reply]
i'm looking into refactoring that <kbd>Cel Shader</kbd> node group into a custom node, since specific node groups can't be created with the Add Node menu directly and it's bugging me

**Reply by enderprism** [reply]
i had some fun making utility node groups

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/f5090661-4bdd-42d7-b79a-f026c0a8ee55" />

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/84353e53-9dae-4e44-a6f5-ed81f154d415" />

i wonder why there aren't ivec and uvec primitives though

<img width="1100" height="509" alt="image" src="https://github.com/user-attachments/assets/90eb5238-483f-444e-8bd2-53cd5f6572b4" />

**Comment by pacrox** [comment]
I've been doing NPR Toon shading sinvce 30years and I must say that, upon experience, MALT is the best render engine, because it produces consinstent flicker-free result.


Two examples I'm working on:
A) procedural city with procedural lines... All the lines are generated in Malt Shader, no texture, no precalculated stuff:
https://github.com/user-attachments/assets/bdcb8fb2-7c06-4719-86c0-038b7831fc6e

B) A "sort-of" lens shader: all the building gets scaled in Z in function of the distance from the camera (it looks like Malt has no camera shaders, unfortunately)
https://github.com/user-attachments/assets/8f1a4b5b-7d7f-443d-9311-c5b895798f20

Here you feel more the effect of this pseudo camera-shader (which is, infact, an object shader):
https://github.com/user-attachments/assets/c6c150c9-9e60-45b0-b6d7-f731955e3475



Don´t stop developing and improving MALT!
Best tool for Expressive Rendering.
